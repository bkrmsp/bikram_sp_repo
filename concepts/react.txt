Component
Types of Component
Function vs Class component
Lifecycle phases

Hooks
    useState
    useEffect
    useDispatch
    useSelector
    useContext
    useReducer

JSX

Virtual DOM
Diffing

How React is better than Other libraries?

Redux
    Architecture
    Action, Reducer

Redux Middleware
    Thunk

React Router
    Browser
    Hash
    Memory
    Router