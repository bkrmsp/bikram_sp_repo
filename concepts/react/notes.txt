* Component
Basic entity in React world.
A function or class that returns HTML in the form of JSX.
Has a Lifecycle.

* JSX
Javascript based XML syntax
Used in React.

* Component Lifecycle phases
Mounting
Update
Unmounting

** Mounting Phase
When component is mounted on DOM i.e. rendered in browser.
When this occurs, it can be detected in function component using useEffect hook with blank Array as second Parameter

** Update Phase
When component is re-rendered/Updated due to 
change in Local State 
Or change in Store State (if subscribed)
Or due to re-render of Parent component that has passed props to child component

When this occurs, it can be detected by using useEffect with dependency passed inside array as second Parameter

** Unmounting Phase
When component is unmounted from DOM i.e. removed from DOM
When this occrus, it can be detected by returning a method from useEffect first parameter i.e. method with blank array as second Parameter


* Example of Phases:

** Mounting/Unmounting Phase:
useEffect(() => {
    //Anything that needs to be executed when Component Mounted
    return () => {
        //Anything that needs to be executed when Component Unmounts
    }
},[]);

** Update Phase:
useEffect(() => {
    //Anything that needs to be executed when Component Updated
},[props.name, age, gender]);

//Here props.name is props changed (received from Parent component)
//Here age is local State
//Here gender is Store state created using useSelector

* Virtual DOM & Diffing Algorithm
Virtual DOM is in-memory representation of actual DOM.
Component creation or updation occurs first in Virtual DOM.
Every element is uniquely identified by a key which is maintained by React.
This helps to identify what part is changed or need to be updated in Virtual DOM.
The purpose of Virtual DOM is to avoid the un-necessary delay or work to re-render the Real DOM.

Diffing Algorithm is used by React to compare the changes b/w Virtual and Real DOM.
Only the different parts are updated in Real DOM and is made in sync with Virtual DOM.

To sum up, React does a process called 
Reconciliation = Render in Virtual DOM + Diffing + Commit (Sync in Real DOM)

React DOM is responsible for updating the Real DOM w.r.t Virtual DOM

When React sees that root element is different then it tears down old DOM and creates new DOM.
This is very much faster than comparing each element.

When React sees root element is same then it compares attributes and look for  differences and then changes.


# Controlled vs Uncontrolled

Controlled Component
BACKGROUND: HTML maintains the values of elements like input, select, textarea etc.
However, in React, we can also control the value and their changes using state and setState methods.
So, if we want a single source of truth, then we go with the latter option and this type of component is called Controlled.

Uncontrolled Component
In this case, we dont want to maintain the changes in values of HTML elements but would need to access them when required.
So, by creating reference to them, we can refer to the element and access the value. However changes wil be controlled by HTML and not react.


# Stateful/Container/Smart vs Stateless/Presentational/Dumb Component
