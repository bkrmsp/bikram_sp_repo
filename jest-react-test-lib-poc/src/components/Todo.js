import React from 'react';


const Todo = ({ todo }) => {
    const h1Ele = <h1>{todo.text}</h1>
    const strikeEle = <strike>{h1Ele}</strike>
    const text = todo.isCompleted ? <strike>{h1Ele}</strike> : h1Ele;
    return (
        <div data-testid={`todo-${ todo.id }`}>
            {text}
        </div>
    )
}

export default Todo;