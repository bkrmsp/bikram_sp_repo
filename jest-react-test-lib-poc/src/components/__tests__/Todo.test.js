import React from 'react';
import { render, screen, cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';

import Todo from '../Todo';

afterEach(() => {
    cleanup();
})
test('should render Completed Todo', () => {
    const todoItem = { id: 1, text: 'Eat food', isCompleted: true };
    render(<Todo todo={todoItem} />);
    const todoElement = screen.getByTestId(`todo-${ todoItem.id }`);
    expect(todoElement).toBeInTheDocument();
    expect(todoElement).toHaveTextContent(`${ todoItem.text }`);
    expect(todoElement).toContainHTML('strike')
});

test('should render Incomplete Todo', () => {
    const todoItem = { id: 2, text: 'Go to sleep', isCompleted: false };
    render(<Todo todo={todoItem} />);
    const todoElement = screen.getByTestId(`todo-${ todoItem.id }`);
    expect(todoElement).toBeInTheDocument();
    expect(todoElement).toHaveTextContent(`${ todoItem.text }`);
    expect(todoElement).not.toContainHTML('strike')
});

test('match snapshot', () => {
    const todoItem = { id: 1, text: 'Eat food', isCompleted: true };
    const tree = renderer.create(<Todo todo={todoItem} />).toJSON();
    expect(tree).toMatchSnapshot(tree);
})