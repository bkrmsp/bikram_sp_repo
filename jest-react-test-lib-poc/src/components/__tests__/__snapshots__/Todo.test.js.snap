// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match snapshot 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            "Eat food",
          ],
          "props": {},
          "type": "h1",
        },
      ],
      "props": {},
      "type": "strike",
    },
  ],
  "props": {
    "data-testid": "todo-1",
  },
  "type": "div",
}
`;
